src/lib/arch/riscv/InstructionMetadata.cc

```diff
@@ -442,6 +449,426 @@ void InstructionMetadata::includeZeroRegisterPosZero() {
+
+  cs_riscv_op temp;
+  temp.type = RISCV_OP_MEM;
+  temp.mem.base = operands[2].reg;
+  temp.mem.disp = operands[1].imm;
+
+  operands[1] = temp;
+
+  operandCount = 2;
+}
+
+void InstructionMetadata::setLength(uint8_t size) {
+  lenBytes = size;
+  switch (size) {
+    case 2:
+      len = INSTR_LENGTH::IL_16B;
+      break;
+    case 4:
+      len = INSTR_LENGTH::IL_32B;
+      break;
+    default:
+      len = INSTR_LENGTH::IL_INVALID;
+  }
+}
+void InstructionMetadata::convertCompressedInstruction(const cs_insn& insn) {
+  if (len != INSTR_LENGTH::IL_16B) {
+    return;
+  }
+
+  switch (insn.opcode) {
+    case Opcode::RISCV_C_JR:
+      // jalr x0, 0(rs1)
+      // trap if rs1 = zero
+      // C.JR Rs, _, _ -> JALR x0, Rs, 0
+      opcode = Opcode::RISCV_JALR;
+
+      operands[0].type = RISCV_OP_REG;
+      operands[0].reg = 1;
+
+      operands[1] = insn.detail->riscv.operands[0];
+
+      operands[2].type = RISCV_OP_IMM;
+      operands[2].imm = 0;
+
+      operandCount = 3;
+
+      break;
+    case Opcode::RISCV_C_MV:
+      // add rd, x0, rs2
+      // trap if rs2 = zero. rs2 == zero corresponds to C.JR
+      // rs2==zero and rd == zero are hints
+      // C.MV rd, rs2, _ -> ADD rd, zero, rs2
+      opcode = Opcode::RISCV_ADD;
+
+      includeZeroRegisterPosOne();
+
+      break;
+    case Opcode::RISCV_C_LDSP: {
+      // TODO valid for RV64 only. Make this check
+      // ld rd, offset[8:3](x2)
+      // offset is immediate scaled by 8
+      // trap if rd=zero
+      opcode = Opcode::RISCV_LD;
+
+      // Create operand formatted like LD instruction
+      createMemOpPosOne();
+
+      break;
+    }
+    case Opcode::RISCV_C_ADDI4SPN:
+      // addi rd ′ , x2, nzuimm[9:2]
+      // trap if nzuimm=zero
+      opcode = Opcode::RISCV_ADDI;
+      // All operands correct
+      break;
+    case Opcode::RISCV_C_LI:
+      // addi rd, x0, imm[5:0]
+      // trap if rd=zero
+      // C.LI rd, imm, _ -> addi rd, zero, imm
+      opcode = Opcode::RISCV_ADDI;
+
+      includeZeroRegisterPosOne();
+
+      break;
+    case Opcode::RISCV_C_ADDI16SP:
+      // Opcode shared with C.LUI but has Rd = x2
+      // addi x2, x2, nzimm[9:4]
+      // trap if nzimm=zero
+      // C.ADDI16SP sp, imm, _ -> addi sp, sp, imm
+      opcode = Opcode::RISCV_ADDI;
+
+      duplicateFirstOp();
+
+      break;
+    case Opcode::RISCV_C_SLLI:
+      // slli rd, rd, shamt[5:0]
+      //
+      // For RV32C, shamt[5] must be zero; the code points with shamt[5]=1 are
+      // reserved for custom extensions. For RV32C and RV64C, the shift amount
+      // must be non-zero; the code points with shamt=0 are HINTs. For all base
+      // ISAs, the code points with rd=x0 are HINTs, except those with
+      // shamt[5]=1 in RV32C.
+
+      // C.SLLI rd, shamt, _ -> slli rd, rd, shamt
+
+      opcode = Opcode::RISCV_SLLI;
+
+      duplicateFirstOp();
+
+      break;
+    case Opcode::RISCV_C_SDSP: {
+      // TODO rv64 ONLY, make check for this
```

Comment:
Not currently. But these instructions use the same encoding space as others when using rv32c. This TODO is a reminder for if/when rv32 is supported to add a check here. This may not be necessary if we are confident capstone is working correctly